goal_src/engine/gfx/lights.gc

```diff
@@ -5,3 +5,97 @@
 ;; name in dgo: lights
 ;; dgos: GAME, ENGINE
 
+(defun light-slerp ((out light) (a light) (b light) (alpha float))
+  "Linearly interpolate between two light's vectors. Alpha is clamped between 0 - 1."
+  (let ((clamped-alpha (fmax 0.0 (fmin 1.0 alpha))))
+   (vector-lerp! (-> out color) (-> a color) (-> b color) clamped-alpha)
+   (vector-deg-slerp
+    (-> out direction)
+    (-> a direction)
+    (-> b direction)
+    clamped-alpha
+    )
+   (let ((f0-2 (-> a levels x))
+         (f1-2 (-> b levels x))
+         )
+    (set! (-> out levels x) (+ f0-2 (* clamped-alpha (- f1-2 f0-2))))
+    )
+   )
+  out
+  )
+
+(defun
+  light-group-slerp
+  "Linearlyl interpolate between each of the 4 vector components of two light-groups.
+   Alpha is clamped between 0 - 1."
+  ((out light-group) (a light-group) (b light-group) (alpha float))
+  (dotimes (vec-index 4)
+   (light-slerp
+    (the-as light (+ (+ (* 48 vec-index) 0) (the-as int out)))
```

Comment:
This is a little bit weird.  It's basically taking an object, converting it to an integer, adding an integer, and converting it back to an object.  This will do the right thing, but probably isn't what they wrote.

To me, this looks like an array access. It's basically doing `(+ object (* index array-stride))`.  In GOAL its possible and somewhat common to have overlapping fields to give you multiple ways to access the same data. For example, in this case, they have 4 lights: `dir0`, `dir1`, `dir2`, `ambi`. My guess is that they also defined an array of 4 lights that overlaps with these, so they could access `dir2` by doing `(-> group dir2)` or by doing `(-> group lights-array 2)`.  They could optionally hide these overlay fields from the inspect methods, so sometimes you have to add it yourself. You could add an array field like this:
```lisp
(deftype light-group (structure)
  ((dir0 light :inline :offset-assert 0)
   (dir1 light :inline :offset-assert 48)
   (dir2 light :inline :offset-assert 96)
   (ambi light :inline :offset-assert 144)
   (lights light 4 :inline :offset 0) ;; this array starts at the same offset as dir0
   )
  :method-count-assert 9
  :size-assert         #xc0
  :flag-assert         #x9000000c0
  )
```

I suspect that the arguments will change from `(the-as light (+ (+ (* 48 vec-index) 0) (the-as int out)))` to `(-> out lights vec-index)`.

Reply:
oh cool, that's pretty interesting.  and yeah the code definitely did not look like a sane person would have wrote it agreed lol